### PR TITLE
feat(code): add function to download notebook outputs

### PR DESCRIPTION
**Pull request by KeijiBranshi**
BUG=b/371574828
CHILD=#185
BLOCKED_BY=go/kaggle-pr/32581,go/kaggle-pr/32632
CC=@rosbo,@dster2,@jplotts 

Extends the same functionality from [`models.py`](https://github.com/Kaggle/kagglehub/blob/8b1fae8632f9d381cebb14ec50c56d7ff5fbeb1b/src/kagglehub/models.py), [`datasets.py`](https://github.com/Kaggle/kagglehub/blob/8b1fae8632f9d381cebb14ec50c56d7ff5fbeb1b/src/kagglehub/datasets.py) and [`competition.py`](https://github.com/Kaggle/kagglehub/blob/8b1fae8632f9d381cebb14ec50c56d7ff5fbeb1b/src/kagglehub/competition.py) to the notebooks at https://kaggle.com/code.

### Changes

[handle.py](https://github.com/Kaggle/kagglehub/pull/184/files#diff-4d86981fdd4a6e41ce621dd3dcafa482e4ca96d5cca7da9b6cbaff147cc3bffb)

- added a new `*Handle` data type
- using `Code` to align with the route on the main site https://kaggle.com/code

[cache.py](https://github.com/Kaggle/kagglehub/pull/184/files#diff-3468a9a96dc65a3a8770b887cdc452e4975b9c934f0376c56a7e39dff7fd778a)

- added new functions to dictate the cached path for notebook outputs based on the properties in `CodeHandle`
- this mostly mirrors the same structure as the `model`, `dataset`, and `competition` paths.
    - the cache structure is split into the _output\_path_, _archive\_path_, a _completion\_file\_marker\_path for individual files in the download payload, and a _completion\_file\_marker\_path for the entire download payload.
- the structure is as follows:
    ```
    <cache_root>/
    └── notebooks/
        └── username/
            └── notebook_slug/
                ├── output.complete  <-- tracker for the entire output
                ├── .complete/       <-- trackers for per file within the output
                │   └── output/
                │       ├── file1.txt.complete
                │       ├── file2.txt.complete
                ├── output.archive   <-- the compressed output (.tar.gz or .zip)
                └── output/          <-- the uncompressed output
                    ├── file1.txt
                    └── file2.txt
    ```

[http_resolver.py](https://github.com/Kaggle/kagglehub/pull/184/files#diff-bac8b2fc0706f6a2f83562279c3095d907c84f377a4f49829bb06935d3e6773a)

- Implemented the `NotebookOutputHttpResolver`
- Note, we don't currently have an API endpoint to download notebook output in a kagglehub-compatible compression format (left a TODO with our internal tracker for this)
- It leverages the our existing `KaggleApiV1Client` + the new cache location mentioned above

[registry.py](https://github.com/Kaggle/kagglehub/pull/184/files#diff-3a93c8bc0f26d8267eb445e7e90f03b2c85c54e8c59e37f997c82104cb5d1541) + [\_\_init\_\_.py](https://github.com/Kaggle/kagglehub/pull/184/files#diff-214a9613a5b623f57fb158c6c784e29e051f95fe72694b446cae592f76b825d8)

- bootstraps the `NotebookOutputHttpResolver` so that it can be called by `kagglehub.notebook_output_download` in `code.py`

[code.py](https://github.com/Kaggle/kagglehub/pull/184/files#diff-e0c454ef6e8643b0efb25013db58209adc53499f2a5a485a8b5bd63cd280899a)

- the entry point to the notebook output downloading functionality
- the file is named `code.py` to align with our navigation paths at https://kaggle.com (similar to `models`, `datasets`, and `competitions`). Open to changing if needed.
- the function is named `notebook_output_download` to be more specific about what's being downoaded

[test_notebook_output_download.py](https://github.com/Kaggle/kagglehub/pull/184/files#diff-04eda2a8f4c98b098d47a0dc61b70c4bfce3a2fba4205abcb9830c69a05001e3)

- integration tests for the new `kagglehub.notebook_output_download` function
- TODO(#185): adding tests in a followup since that requires propping up a [stubbed API server](https://github.com/Kaggle/kagglehub/tree/8b1fae8632f9d381cebb14ec50c56d7ff5fbeb1b/tests/server_stubs). Trying to keep this diff from getting any bigger.

[gcs_upload.py](https://github.com/Kaggle/kagglehub/pull/184/files#diff-2dcd4fa7b008e12dc0d76354035bc2cdaf48dd2422e997476443a4fbe550d8ea)

- A miscellaneous lint error that slipped through. Fixing here as a drive-by change as per [this comment](https://github.com/Kaggle/kagglehub/pull/184#issuecomment-2494710160)

### CODE REVIEW

**Comment by KeijiBranshi** [comment]
Not sure why this build error is happening:

```
Step #3 - "lint": cmd [1] | ruff check .
Step #3 - "lint": src/kagglehub/gcs_upload.py:13:33: A004 Import `ConnectionError` is shadowing a Python builtin
Step #3 - "lint":    |
Step #3 - "lint": 12 | import requests
Step #3 - "lint": 13 | from requests.exceptions import ConnectionError, Timeout
Step #3 - "lint":    |                                 ^^^^^^^^^^^^^^^ A004
Step #3 - "lint": 14 | from tqdm import tqdm
Step #3 - "lint": 15 | from tqdm.utils import CallbackIOWrapper
Step #3 - "lint":    |
Step #3 - "lint": 
Step #3 - "lint": src/kagglehub/gcs_upload.py:69:16: UP031 Use format specifiers instead of percent format
Step #3 - "lint":    |
Step #3 - "lint": 67 |             suffix_index += 1
Step #3 - "lint": 68 |             size /= 1024.0
Step #3 - "lint": 69 |         return "%.*f%s" % (precision, size, suffixes[suffix_index])
Step #3 - "lint":    |                ^^^^^^^^ UP031
Step #3 - "lint":    |
Step #3 - "lint":    = help: Replace with format specifiers
Step #3 - "lint": 
Step #3 - "lint": Found 2 errors.
```

that file doesn't appear in this diff at all. cc @rosbo in case you know

**Comment by rosbo** [comment]
@KeijiBranshi ruff is not pinned to a specific version. This is probably a new check: https://github.com/Kaggle/kagglehub/blob/8b1fae8632f9d381cebb14ec50c56d7ff5fbeb1b/pyproject.toml#L58

`ruff` 0.8.0 was released 10 hours ago: https://pypi.org/project/ruff/#history

We could pin `ruff` but at the same time, we would need to remember to bump the version up. If it's only a single error, you can simply fix it. If you have many error, just pin ruff to 0.7.4 and file a bug to bump up the version.

File: src/kagglehub/handle.py [hunk 1]
```diff
@@ -83,6 +83,25 @@ def to_url(self) -> str:
...
+class CodeHandle(ResourceHandle):
+    owner: str
+    notebook: str
+    version: Optional[int] = None
+
+    def is_versioned(self) -> bool:
+        return self.version is not None and self.version > 0
+
+    def __str__(self) -> str:
+        handle_str = f"{self.owner}/{self.notebook}"
```
Comment (rosbo) [inline src/kagglehub/handle.py:96]: You should include the version in the __str__ and to_url if it the user specifies a specific version of the kernel.

Comment (rosbo) [reply src/kagglehub/handle.py:96]: I see the parse method below doesn't support versioned handles, are you planning to support it?

If not, you can remove `version` and `is_versioned` from `CodeHandle`.

**Review by rosbo — CHANGES_REQUESTED** [review]
You can also add an integration test (won't need a stub server) now and should be fairly simple.

Example: https://github.com/Kaggle/kagglehub/blob/8b1fae8632f9d381cebb14ec50c56d7ff5fbeb1b/integration_tests/test_dataset_download.py#L14-L33

Instructions to run the integration tests: https://github.com/Kaggle/kagglehub/tree/main?tab=readme-ov-file#integration-tests

File: src/kagglehub/cache.py [hunk 2]
```diff
@@ -4,9 +4,10 @@
 from typing import Optional
 
 from kagglehub.config import get_cache_folder
-from kagglehub.handle import CompetitionHandle, DatasetHandle, ModelHandle, ResourceHandle
+from kagglehub.handle import CodeHandle, CompetitionHandle, DatasetHandle, ModelHandle, ResourceHandle
```
Comment (calderjo) [inline src/kagglehub/cache.py:7]: I think we should stick with the term notebook here to be consistent. 

and ditto in other areas.

File: src/kagglehub/http_resolver.py [hunk 3]
```diff
@@ -199,6 +200,66 @@ def _inner_download_file(file: str) -> None:
...
+        thread_map(
+            _inner_download_file,
+            files,
+            desc=f"Downloading {len(files)} files",
+            max_workers=8,  # Never use more than 8 threads in parallel to download files.
+        )
+
+        mark_as_complete(h, path)
+
+        # TODO(b/377510971): when notebook is a Kaggle utility script, update sys.path
```
Comment (rosbo) [inline src/kagglehub/http_resolver.py:245]: When it is a utility script, the source and not the output is used. Is the name `notebook_output_download` still the right name?

File: src/kagglehub/cache.py (hunk 2 above)
Comment (rosbo) [reply src/kagglehub/cache.py:7]: We could also use Kernel (which is what we use usually in our codebase).

File: src/kagglehub/http_resolver.py [hunk 4]
```diff
@@ -199,6 +200,66 @@ def _inner_download_file(file: str) -> None:
...
+        api_path = f"kernels/{h.owner}/{h.notebook}/output/download"
+        output_root = Path(get_cached_path(h, path))
+
+        # Create the intermediary directories
+        # List the files and decide how to download them:
+        # - <= 25 files: Download files in parallel
+        # > 25 files: Download the archive and uncompress
+        (files, has_more) = self._list_files(api_client, h)
+        if has_more:
+            # TODO(b/379761520): support .tar.gz archived downloads
```
Comment (calderjo) [inline src/kagglehub/http_resolver.py:226]: maybe misplaced, are .tar.gz files supposed to fail at this point?

File: src/kagglehub/http_resolver.py (hunk 3 above)
Comment (KeijiBranshi) [reply src/kagglehub/http_resolver.py:245]: but maybe cc @djherbis @calderjo to confirm

File: src/kagglehub/http_resolver.py (hunk 4 above)
Comment (KeijiBranshi) [reply src/kagglehub/http_resolver.py:226]: yeah, this is how `model_download` knows when to download an archive vs individual files:

https://github.com/Kaggle/kagglehub/blob/8b1fae8632f9d381cebb14ec50c56d7ff5fbeb1b/src/kagglehub/http_resolver.py#L168-L172

But we don't have an API endpoint that returns the supported compression formats. See the bug for more context.

File: src/kagglehub/cache.py (hunk 2 above)
Comment (calderjo) [reply src/kagglehub/cache.py:7]: kernel is used internally, notebook is the "public term".
i think since this is public repo, we should stick to notebook. 

using it interchangeably in the same repo, will get pretty confusing for both us and external users.

Comment (KeijiBranshi) [reply src/kagglehub/cache.py:7]: following that distinction, should the public api endpoint also fall under `/api/v1/notebooks/`?

File: src/kagglehub/http_resolver.py (hunk 3 above)
Comment (calderjo) [reply src/kagglehub/http_resolver.py:245]: We do convert ipynb to a py and place it in both the source and output directories ([link](https://source.corp.google.com/h/kaggle-internal/codesearch/+/master:kaggle/Kaggle.Services.Kernels/Clients/SessionSchedulerClient.cs;l=434-455?q=%22utility%22%20%22scripts%22%20%22output%22&ss=h%2Fkaggle-internal%2Fcodesearch%2F%2B%2Frefs%2Fheads%2Fmaster:kaggle%2FKaggle.Services.Kernels%2F))

File: src/kagglehub/cache.py (hunk 2 above)
Comment (calderjo) [reply src/kagglehub/cache.py:7]: public api endpoint should remain as kernel, i meant code written in kagglehub like "CodeHandle" and "CODE_CACHE_SUBFOLDER"

File: src/kagglehub/http_resolver.py (hunk 3 above)
Comment (rosbo) [reply src/kagglehub/http_resolver.py:245]: Great. Thanks for checking.

**Comment by KeijiBranshi** [comment]
> @KeijiBranshi ruff is not pinned to a specific version. This is probably a new check:
> 
> https://github.com/Kaggle/kagglehub/blob/8b1fae8632f9d381cebb14ec50c56d7ff5fbeb1b/pyproject.toml#L58
> 
> `ruff` 0.8.0 was released 10 hours ago: https://pypi.org/project/ruff/#history
> 
> We could pin `ruff` but at the same time, we would need to remember to bump the version up. If it's only a single error, you can simply fix it. If you have many error, just pin ruff to 0.7.4 and file a bug to bump up the version.

Fixed here 0cd1661d43a2d7c221dd49570c693b236f019696

File: src/kagglehub/cache.py (hunk 2 above)
Comment (KeijiBranshi) [reply src/kagglehub/cache.py]: updated everything to say "notebooks" 1876af0eebe9f987dc265f17339289ee8cd9d31b

File: src/kagglehub/gcs_upload.py [hunk 5]
```diff
@@ -10,7 +10,7 @@
 from typing import Optional, Union
 
 import requests
-from requests.exceptions import ConnectionError, Timeout
+from requests.exceptions import Timeout
```
Comment (KeijiBranshi) [inline src/kagglehub/gcs_upload.py:13]: note: these are miscellaneous drive-by changes as per #issuecomment-2494710160

**Review by rosbo — APPROVED** [review]

**Comment by KeijiBranshi** [comment]
fyi @calderjo going to merge now to unblock. lmk if you have any comments though and I can address in a follow up iteration
